bokeh/models/tests/test_sources.py

```diff
@@ -148,5 +148,21 @@ def mock(*args, **kw):
         self.assertEqual(stuff['args'], ("doc", ds, dict(a=[(0,100), (1,101)], b=[(0,200)])))
         self.assertEqual(stuff['kw'], {})
 
+    def test_data_column_lengths(self):
+        with self.assertRaises(ValueError):
+            ColumnDataSource(data=dict(a=[10, 11], b=[20, 21, 22]))
+
+        ds = ColumnDataSource()
+        with self.assertRaises(ValueError):
+            ds.data = dict(a=[10, 11], b=[20, 21, 22])
+
+        ds = ColumnDataSource(data=dict(a=[10, 11]))
+        with self.assertRaises(ValueError):
+            ds.data["b"] = [20, 21, 22]
+
+        ds = ColumnDataSource(data=dict(a=[10, 11], b=[20, 21]))
+        with self.assertRaises(ValueError):
+            ds.data.update(dict(a=[10, 11, 12]))
```

Comment:
Note that with the current implementation of property containers, `ds.data` is updated prior to assertions being check, so its value is still mutated, which can be confusing (e.g. in an interactive session). I think we can do something about this, but I'm not sure how expensive this will be.